provider/pkg/kinds/deprecated.go

```diff
@@ -66,48 +67,146 @@ import (
+var v119 = cluster.ServerVersion{Major: 1, Minor: 19}
+var v120 = cluster.ServerVersion{Major: 1, Minor: 20}
+var v122 = cluster.ServerVersion{Major: 1, Minor: 22}
+
 func gvkStr(gvk schema.GroupVersionKind) string {
 	return gvk.GroupVersion().String() + "/" + gvk.Kind
 }
 
-// DeprecatedAPIVersion returns true if the given GVK is deprecated in the most recent k8s release.
-func DeprecatedAPIVersion(gvk schema.GroupVersionKind) bool {
-	return SuggestedAPIVersion(gvk) != gvkStr(gvk)
+func gvkFromStr(str string) schema.GroupVersionKind {
+	parts := strings.Split(str, "/")
+	return schema.GroupVersionKind{Group: parts[0], Version: parts[1], Kind: parts[2]}
+}
+
+// DeprecatedAPIVersion returns true if the given GVK is deprecated in the given k8s release.
+func DeprecatedAPIVersion(gvk schema.GroupVersionKind, version *cluster.ServerVersion) bool {
+	suggestedGVKString := SuggestedAPIVersion(gvk)
+	if version == nil {
+		// If no version is provided, check only against the latest version.
+		return suggestedGVKString != gvkStr(gvk)
+	}
+
+	suggestedGVK := gvkFromStr(suggestedGVKString)
+	suggestedGVKExists := ExistsInVersion(suggestedGVK, version)
+
+	return suggestedGVKExists && suggestedGVKString != gvkStr(gvk)
+}
+
+// AddedInVersion returns the ServerVersion of k8s that a GVK is added in.
+func AddedInVersion(gvk schema.GroupVersionKind) *cluster.ServerVersion {
+	gv, k := groupVersion(gvk.GroupVersion().String()), Kind(gvk.Kind)
+
+	switch gv {
+	case AdmissionregistrationV1:
+		switch k {
+		case MutatingWebhookConfiguration, MutatingWebhookConfigurationList, ValidatingWebhookConfiguration, ValidatingWebhookConfigurationList:
+			return &v116
+		}
+	case ApiextensionsV1:
+		switch k {
+		case CustomResourceDefinition, CustomResourceDefinitionList:
+			return &v116
+		}
+	case AuthenticationV1:
+		switch k {
+		case TokenRequest:
+			return &v116
+		}
+	case CoordinationV1:
+		switch k {
+		case Lease, LeaseList:
+			return &v114
+		}
+	case DiscoveryV1B1:
+		switch k {
+		case EndpointSlice, EndpointSliceList:
+			return &v117
+		}
+	case FlowcontrolV1A1:
+		switch k {
+		case FlowSchema, FlowSchemaList, PriorityLevelConfiguration, PriorityLevelConfigurationList:
+			return &v117
+		}
+	case NetworkingV1B1:
+		switch k {
+		case Ingress, IngressList:
+			return &v114
+		case IngressClass, IngressClassList:
+			return &v118
+		}
+	case NodeV1A1:
+		switch k {
+		case RuntimeClass, RuntimeClassList:
+			return &v114
+		}
+	case NodeV1B1:
+		switch k {
+		case RuntimeClass, RuntimeClassList:
+			return &v114
+		}
+	case SchedulingV1:
+		switch k {
+		case PriorityClass, PriorityClassList:
+			return &v114
+		}
+	case StorageV1B1:
+		switch k {
+		case CSIDriver, CSIDriverList, CSINode, CSINodeList:
+			return &v114
+		}
+	case StorageV1:
+		switch k {
+		case CSINode, CSINodeList:
+			return &v117
+		case CSIDriver, CSIDriverList:
+			return &v118
+		}
+	}
+
+	return &cluster.ServerVersion{Major: 1, Minor: 0}
```

Comment:
It looks like this is only checking back to v1.14, so this is not quite correct. Can you extend this logic back to v1.10? That would cover the last two years of releases, which should account for the majority of cases.

Reply:
Sounds good will do.